File: kotlin/src/main/kotlin/Projectile.kt
fun main() {
    println("~~~~~Starting~~~~~")

    val gravity = Vector(0.0, -0.1, 0.0)
    val wind = Vector(-0.01, 0.0, 0.0)
    val environment = Pair(gravity, wind)
    val projectile = Pair(Point(0.0, 0.0, 0.0), Vector(1.0, 1.0, 0.0).normalize())

    val points = iterate(listOf(projectile), environment)

    val width = 200
    val height = 200
    val red = Color(1.0, 0.0, 0.0)
    val canvas = Canvas(width, height)
    points.map { it.first }
        .map { point -> Pair((point.x * 10).toInt(), (point.y * 10).toInt())}
        .map { (x, y) -> Pair(x, height - y - 1) }
        .forEach { (x, y) -> canvas.writePixel(x, y, red)}

    writePpm(canvasToPpm(canvas))

    println("~~~~~Done~~~~~")
}

fun iterate(path: List<Pair<Point, Vector>>, environment: Pair<Vector, Vector>): List<Pair<Point, Vector>> {
    val (location, velocity) = path.last()
    val (gravity, wind) = environment
    println(location)

    val newLocation = location.plus(velocity)
    if (newLocation.y < 0) return path

    val newVelocity = velocity.plus(gravity).plus(wind)
    return iterate(path.plus(Pair(newLocation, newVelocity)), environment)
}

File: rust/src/exercises/projectile.rs
use crate::display::canvas::Canvas;
use crate::display::color::Color;
use crate::display::writer::write_canvas;
use crate::geometry::point::Point;
use crate::geometry::vector::Vector;

pub fn run() {
    let gravity = Vector::new(0.0, -0.1, 0.0);
    let wind = Vector::new(-0.01, 0.0, 0.0);
    let environment = (gravity, wind);
    let projectile = (
        Point::at(0.0, 0.0, 0.0),
        Vector::new(1.0, 1.0, 0.0).normalize(),
    );

    let points = iterate(vec![projectile], environment);

    let width = 200;
    let height = 200;
    let mut canvas = Canvas::new(width, height);
    points
        .iter()
        .map(|(point, _)| *point)
        .map(|point| ((point.x * 10.0) as usize, (point.y * 10.0) as usize))
        .map(|(x, y)| (x, height - y - 1))
        .for_each(|(x, y)| canvas.write_pixel(x, y, Color::RED));

    write_canvas(&canvas).unwrap();
}

fn iterate(mut path: Vec<(Point, Vector)>, environment: (Vector, Vector)) -> Vec<(Point, Vector)> {
    let (location, velocity) = path.last().unwrap();
    let (gravity, wind) = environment;

    let new_location = *location + *velocity;
    if (new_location.y < 0.0) {
        return path;
    }

    let new_velocity = *velocity + gravity + wind;
    path.push((new_location, new_velocity));
    iterate(path, environment)
}
